### DISCUSSION TITLE
Curious about implementing deallocators for buffer returned from Rust

### DISCUSSION THREAD

**Discussion by hsfzxjy** [fzyzcjy/flutter_rust_bridge#718] [Q&A] [answered]
Assume I am writing an API where a heap-allocated byte buffer is returned from Rust to Dart side. The byte buffer will be processed by some Dart code, and then it should be deallocated. 

I was wondering if there's any approach to automatically do the deallocation job by, for example, calling a Rust deallocator function when the Dart object is about to be GCed, or I must manually call the deallocator by myself?

**Comment by fzyzcjy** [comment, marked as answer]
> if there's any approach to automatically do the deallocation job by, for example, calling a Rust deallocator function when the Dart object is about to be GCed, or I must manually call the deallocator by myself?

If you say `ZeroCopyBuffer<Vec<u8>>`, then it is auto deallocated when Dart decides to dispose it. Details can be found in `allo-isolate`, which is a wrapper around Dart's C api.

if you say any arbitrary data, they are copied (instead of zero-copy) to Flutter, and it is still Flutter's job to deallocate.

**Comment by hsfzxjy** [comment]
@fzyzcjy Thanks for your reply!

btw, is it possible to implement zero-copied arbitary struct that can also be auto-deallocated? Currently only typed arrays support this feature. 

I've spotted there's `Dart_CObject.value.as_native_pointer` in [dart_native_api.h](https://github.com/dart-lang/sdk/blob/main/runtime/include/dart_native_api.h#L92) , which seemingly wraps a struct pointer along with a native deallocator. May this be used for the purpose?

**Reply by fzyzcjy** [reply]
> implement zero-copied arbitary struct that can also be auto-deallocated
I've spotted there's Dart_CObject.value.as_native_pointer in [dart_native_api.h](https://github.com/dart-lang/sdk/blob/main/runtime/include/dart_native_api.h#L92) , which seemingly wraps a struct pointer along with a native deallocator. May this be used for the purpose?

Seems no. We have had some discussions with the dart team and realized impossible. You may search in this repo and dart repo, and possibly also https://github.com/fzyzcjy/flutter_rust_bridge/pull/293.

**Reply by hsfzxjy** [reply]
Many thanks again for the instruction.

**Reply by fzyzcjy** [reply]
You are welcome!
